src/opentracing-c/span.h

```diff
@@ -0,0 +1,252 @@
+    /** Time span finished using monotonic clock. */
+    opentracing_duration finish_time;
+
+    /** Array of log records. */
+    const opentracing_log_record* log_records;
+
+    /**
+     * Number of log records. If log_records is NULL, num_log_records must be
+     * set to zero.
+     */
+    int num_log_records;
+} opentracing_finish_span_options;
+
+/* Forward declaration. */
+struct opentracing_tracer;
+
+/**
+ * Span interface.
+ * @extends opentracing_destructible
+ */
+typedef struct opentracing_span {
+    /** Base class member. */
+    opentracing_destructible base;
+
+    /**
+     * Sets the end timestamp and finalizes span state.
+     *
+     * With the exception of calls to context (which are always allowed),
+     * finish must be the last call made to any span instance, and to do
+     * otherwise leads to undefined behavior.
+     *
+     * @param span Span instance.
+     * @see context
+     */
+    void (*finish)(struct opentracing_span* span);
+
+    /**
+     * Like finish() but with explicit control over timestamps and log data.
+     * @param span Span instance.
+     * @param options Options to override in span completion.
+     * @see finish
+     */
+    void (*finish_with_options)(struct opentracing_span* span,
+                                const opentracing_finish_span_options* options);
+
+    /**
+     * Yields the opentracing_span_context for this span. Note that the
+     * return value of context is still valid after a call to finish, as is
+     * a call to context after a call to finish.
+     * @param span Span instance.
+     * @return Span context associated with this span.
+     * @see finish
+     */
+    opentracing_span_context* (*span_context)(struct opentracing_span* span);
+
+    /**
+     * Sets or changes the operation name.
+     * @param span Span instance.
+     * @param operation_name New operation name.
+     * @return opentracing_true if successful, opentracing_false otherwise
+     *         (indicates out of memory).
+     */
+    opentracing_bool (*set_operation_name)(struct opentracing_span* span,
+                                           const char* operation_name);
+
+    /**
+     * Adds a tag to the span. If there is a pre-existing tag set for `key`,
+     * it is overwritten. Tag values can be numeric types, strings, or bools.
+     * The behavior of other tag value types is undefined at the OpenTracing
+     * level. If a tracing system does not know how to handle a particular
+     * value type, it may ignore the tag, but shall not panic.
+     * @param span Span instance.
+     * @param key Tag key.
+     * @param value Tag value.
+     * @return opentracing_true if successful, opentracing_false otherwise
+     *         (indicates out of memory).
+     */
+    opentracing_bool (*set_tag)(struct opentracing_span* span,
+                                const char* key,
+                                const opentracing_value* value);
```

Comment:
Do we want to be returning an error code for these functions? It doesn't seem like there's anything a user would want to do other than log, which a tracer implementation would be doing anyways (the [Go API](https://github.com/opentracing/opentracing-go/blob/master/span.go#L54) doesn't use error codes).

Reply:
I guess I agree we can avoid burdening the user with these issues. If you don't think there is a case where a user needs to know that `set_tag` etc. succeeded or OOM occurred, then I don't have a problem using `void` as the return type.